src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -0,0 +1,431 @@
+			 * instead of using parallel connections.
+			 */
+			scanState->customScanState.methods = &RouterSequentialModifyCustomExecMethods;
+		}
+		else
+		{
+			scanState->customScanState.methods = &RouterMultiModifyCustomExecMethods;
+		}
+	}
+
+	return (Node *) scanState;
+}
+
+
+/*
+ * CoordinatorInsertSelectCrateScan creates the scan state for executing
+ * INSERT..SELECT into a distributed table via the coordinator.
+ */
+static Node *
+CoordinatorInsertSelectCreateScan(CustomScan *scan)
+{
+	CitusScanState *scanState = palloc0(sizeof(CitusScanState));
+
+	scanState->executorType = MULTI_EXECUTOR_COORDINATOR_INSERT_SELECT;
+	scanState->customScanState.ss.ps.type = T_CustomScanState;
+	scanState->distributedPlan = GetDistributedPlan(scan);
+
+	scanState->customScanState.methods = &CoordinatorInsertSelectCustomExecMethods;
+
+	return (Node *) scanState;
+}
+
+
+/*
+ * DelayedErrorCreateScan is only called if we could not plan for the given
+ * query. This is the case when a plan is not ready for execution because
+ * CreateDistributedPlan() couldn't find a plan due to unresolved prepared
+ * statement parameters, but didn't error out, because we expect custom plans
+ * to come to our rescue. But sql (not plpgsql) functions unfortunately don't
+ * go through a codepath supporting custom plans. Here, we error out with this
+ * delayed error message.
+ */
+static Node *
+DelayedErrorCreateScan(CustomScan *scan)
+{
+	DistributedPlan *distributedPlan = GetDistributedPlan(scan);
+
+	/* raise the deferred error */
+	RaiseDeferredError(distributedPlan->planningError, ERROR);
+
+	return NULL;
+}
+
+
+/*
+ * CitusSelectBeginScan is an empty function for BeginCustomScan callback.
+ */
+static void
+CitusSelectBeginScan(CustomScanState *node, EState *estate, int eflags)
+{
+	/* just an empty function */
+}
+
+
+/*
+ * CitusEndScan is used to clean up tuple store of the given custom scan state.
+ */
+static void
+CitusEndScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+
+	if (scanState->tuplestorestate)
+	{
+		tuplestore_end(scanState->tuplestorestate);
+		scanState->tuplestorestate = NULL;
+	}
+}
+
+
+/*
+ * CitusReScan is just a place holder for rescan callback. Currently, we don't
+ * support rescan given that there is not any way to reach this code path.
+ */
+static void
+CitusReScan(CustomScanState *node)
+{
+	ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					errmsg("rescan is unsupported"),
+					errdetail("We don't expect this code path to be executed.")));
+}
+
+
+/*
+ * Load data collected by real-time or task-tracker executors into the tuplestore
+ * of CitusScanState. For that, we first create a tuple store, and then copy the
+ * files one-by-one into the tuple store.
+ *
+ * Note that in the long term it'd be a lot better if Multi*Execute() directly
+ * filled the tuplestores, but that's a fair bit of work.
+ */
+void
+LoadTuplesIntoTupleStore(CitusScanState *citusScanState, Job *workerJob)
```

Comment:
Does this function really belong here? It seems more of a generic executor utility function  (multi_server_executor?).

For example, we'll use part of this logic for the `read_records_file` function.

Reply:
Makes sense, moving to `multi_executor` because there is an interdependency among`multi_server_executor ` and `citus_custom_scan`